cpp/unit/producer.cpp

```diff
@@ -100,211 +235,488 @@ bool UnitTypeTest::place(Unit *unit, Terrain *terrain, coord::tile init_tile) {
-	if (not (tex_die && tex_stand && tex_walk && tex_attack)) {
-		return nullptr;
+std::string BuldingProducer::producer_name() const {
+	return this->unit_data.name;
+}
+
+void BuldingProducer::initialise(Unit *unit, Player &player) {
+	log::dbg("building %lu added with type id %u", unit->id, this->unit_data.id0);
+	log::dbg("standing graphic = %u", this->unit_data.graphic_standing0);
+
+	// initialize graphic set
+	unit->producer = this;
+	unit->unit_class = this->unit_data.unit_class;
+	unit->graphics = &this->graphics;
+
+	auto player_attr = new Attribute<attr_type::owner>(player);
+	unit->add_attribute(player_attr);
+	unit->add_attribute(new Attribute<attr_type::building>());
+	unit->add_attribute(new Attribute<attr_type::garrison>());
+	unit->add_attribute(new Attribute<attr_type::hitpoints>(this->unit_data.hit_points));
+
+	// basic attributes on annexes
+	for (auto obj : unit->location->get_children()) {
+		obj->unit.add_attribute(player_attr);
+		obj->unit.push_action(std::make_unique<IdleAction>(&obj->unit), true);
 	}
 
-	// get required sounds
-	TestSound *snd_create  = this->main->find_sound(unit.sound_creation0);
-	TestSound *snd_destroy = this->main->find_sound(unit.sound_dying);
-	TestSound *snd_move    = this->main->find_sound(unit.move_sound);
+	auto &bl_attr = unit->get_attribute<attr_type::building>();
+	bl_attr.pp = trainable2;
+
+	// if a destroyed texture exists
+	if (this->destroyed) {
+		unit->push_action(std::make_unique<DeadAction>(unit), true);
+	}
+	unit->push_action(std::make_unique<IdleAction>(unit), true);
 
-	return std::make_unique<UnitTypeTest>(
-		&unit,
-		tex_die, tex_stand, tex_walk, tex_attack,
-		snd_create, snd_destroy, snd_move, snd_move
-	);
+	if (this->unit_data.projectile_unit_id > 0 && this->projectile) {
+		coord::phys_t range_phys = coord::settings::phys_per_tile * this->unit_data.max_range;
+		unit->add_attribute(new Attribute<attr_type::attack>(this->projectile, range_phys, 350000, 1, this->graphics));
+		unit->give_ability(std::make_shared<AttackAbility>());
+	}
+
+	// building can train new units and ungarrison
+	unit->give_ability(std::make_shared<TrainAbility>());
+	unit->give_ability(std::make_shared<UngarrisonAbility>());
 }
 
-std::unique_ptr<UnitProducer> ProducerLoader::load_object(const gamedata::unit_object &object) {
-	Texture *tex = this->main->find_graphic(object.graphic_standing0);
-	if ( !tex ) {
+TerrainObject *BuldingProducer::place(Unit *u, Terrain &terrain, coord::phys3 init_pos) {
+
+	// not a tc --- hacks / fix me
+	bool collisions = this->unit_data.id0 != 109;
+
+	// buildings have a square base
+	TerrainObject *obj = new SquareObject(*u, this->foundation_size, this->terrain_outline, collisions);
+
+	/*
+	 * decide what terrain is passable using this lambda
+	 * currently unit avoids water and tiles with another unit
+	 * this function should be true if pos is a valid position of the object
+	 */
+	obj->passable = [obj, &terrain](const coord::phys3 &pos) -> bool {
```

Comment:
we should figure out the terrainrestrictions to get this information dynamically.